tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReTLALanguagePred.scala

```diff
@@ -0,0 +1,179 @@
+      // get expanded
+      case OperEx(TlaFunOper.app, fn, arg) =>
+        isOkInContext(letDefs, fn).and(
+            arg match {
+              case OperEx(TlaFunOper.tuple, args @ _*) => allArgsOk(letDefs, args)
+              case _                                   => isOkInContext(letDefs, arg)
+            },
+        )
+
+      // SANY quirk #19123:
+      // [f EXCEPT ![1] = ...] expands to
+      // OperEx( except, << 1 >>, ... ) and
+      // [f EXCEPT ![1,2] = ...] expands to
+      // OperEx(except, << << 1, 2 >> >> , ...)
+      // reTLA supports ONLY one-at-a-time, single-level EXCEPT, i.e.
+      // [f EXCEPT ![a,b,c,...] = r] (not [f EXCEPT ![a][b] = r] or [f EXCEPT ![a] = r, ![b] = s] )
+      case OperEx(TlaFunOper.except, fn, key, value) =>
+        isOkInContext(letDefs, fn)
+          .and(
+              isOkInContext(letDefs, value),
+          )
+          .and(
+              key match {
+                // ![a,b,...] case
+                case OperEx(TlaFunOper.tuple, OperEx(TlaFunOper.tuple, args @ _*)) =>
+                  allArgsOk(letDefs, args)
+                // ![a] case
+                case OperEx(TlaFunOper.tuple, args @ _*) =>
+                  allArgsOk(letDefs, args)
+                // Impossible, but we need case-completeness
+                case _ => PredResultFail(List())
+              },
+          )
+
+      case LetInEx(body, defs @ _*) =>
+        // go inside the let definitions (similar to FlatLanguagePred)
+        def eachDefRec(ctx: Set[String], ds: List[TlaOperDecl]): PredResult = {
+          ds match {
+            case Nil =>
+              PredResultOk()
+
+            case head :: tail =>
+              isOkInContext(ctx, head.body) // check the first operator definition
+                .and(eachDefRec(ctx + head.name, tail)) // check the other operator definitions
+          }
+        }
+
+        // check the let-definitions first, in a sequence, as they may refer to each other
+        val defsResult = eachDefRec(letDefs, defs.toList)
+        val newLetDefs = defs.map(_.name).toSet
+        // check the terminal expression in the LET-IN chain, by assuming the context generated by the definitions
+        defsResult
+          .and(isOkInContext(letDefs ++ newLetDefs, body))
+
+      case e @ OperEx(TlaOper.apply, NameEx(opName), args @ _*) =>
+        // the only allowed case is calling a nullary operator that was declared with let-in
+        if (!letDefs.contains(opName)) {
+          PredResultFail(List((e.ID, s"undeclared operator $opName")))
+        } else if (args.nonEmpty) {
+          PredResultFail(List((e.ID, s"non-nullary operator $opName")))
+        } else {
+          PredResultOk()
+        }
+
+      case e =>
+        PredResultFail(List((e.ID, e.toString)))
+    }
+  }
+}
+
+object ReTLALanguagePred {
+  private val singleton = new ReTLALanguagePred
+
+  protected val unaryOps: HashSet[TlaOper] =
```

Comment:
I know this structure wasn't introduced in your code here, but do you know what is the motive for having these distinct `HashSets` rather than one `HashSet` (or, equivalently, a predicate `TlaOper -> Boolean`)?  Is it a kind of micro optimization? Or are these subsets used in some other way I'm missing?

Reply:
You know, I honestly hadn't thought about it. Looking at it again it does seem unnecessary, fold should work in all cases, even for `bindingOps`.